blog/content/post/2021-10-28-rockchip-secure-boot.md

```diff
@@ -0,0 +1,375 @@
+The most interesting is the last section called `Efuse power up`.
+
+![eFUSE power up](/img/rk_efuse_power_up.png)
+
+I downloaded board schematic from the
+[Firefly support page](https://en.t-firefly.com/doc/download/4.html) (note that
+there are two board versions). Each download is a zip archive containing a
+schematic and board component layout. I didn't know which board revision I had.
+Also, I couldn't see any differences in schematics or layouts, so I
+double-checked that the relevant part is the same on both revisions.
+
+This is what I have found:
+
+![eFUSE test point](/img/firefly_rk3288_efuse_testpoint.png)
+
+EFUSE_VQPS pin is connected to test point T17, located on the backside, behind
+the SoC.
+
+![T17 location](/img/firefly_t17_location.png)
+
+On the board, it is located here.
+
+![T17 location on board](/img/firefly_t17_location_on_board.jpg)
+
+### Attempt 2
+
+I took a lab power supply. I tried to find some thin probes without success, so
+I used the probes I had, together with male jumper wires.
+
+![powering eFUSEs](/img/powering_efuses.jpg)
+
+It took me a few tries before I succeeded - I was touching a tiny resistor
+placed right next to the test point, draining too much power, around 300 mA.
+This is very dangerous for the board. Luckily, my board survived. After a few
+tries, I got it right, and eFUSE burning succeeded.
+
+![eFUSE burn successful](/img/efuse_tool_success.png)
+
+## What's next
+
+Board got into MaskROM mode, and I successfully ran a signed loader using
+`rkdeveloptool db`. Messages from the serial console show that secure mode is
+active.
+
+```
+DDR Version 1.09 20201119
+In
+Channel a: DDR3 400MHz
+Bus Width=32 Col=10 Bank=8 Row=15 CS=1 Die Bus-Width=16 Size=1024MB
+Channel b: DDR3 400MHz
+Bus Width=32 Col=10 Bank=8 Row=15 CS=1 Die Bus-Width=16 Size=1024MB
+Memory OK
+Memory OK
+OUT
+Boot1 Release Time: Nov 27 2019 15:30:08, version: 2.58
+ChipType = 0x8, 248
+mmc2:cmd19,100
+SdmmcInit=2 0
+BootCapSize=1000
+UserCapSize=14910MB
+FwPartOffset=2000 , 1000
+mmc0:cmd5,20
+SdmmcInit=0 0
+BootCapSize=0
+UserCapSize=30436MB
+FwPartOffset=2000 , 0
+StorageInit ok = 46317
+SecureMode = 1
+SecureInit read PBA: 0x4
+atags_set_pub_key: ret:(0)
+SecureInit ret = 0, SecureMode = 1
+(...)
+```
+
+Loading of unsigned one no longer works - `rkdeveloptool` hangs. MaskROM won't
+accept any commands until DDR Init is loaded, thus preventing unauthorized
+access to eMMC and data dump. U-Boot from SD card no longer boots. You can see
+it for yourself on the following video:
+
+[![asciicast](https://asciinema.org/a/c4J0pRqWL4WlYeqVH04MGLSiC.svg)](https://asciinema.org/a/c4J0pRqWL4WlYeqVH04MGLSiC)
+
+Loader can be flashed onto eMMC using the following commands. Note that this
+flashes only loader, without U-Boot itself.
+
+```
+rkdeveloptool db RK3288_Loader_signed.bin
+rkdeveloptool ul RK3288_Loader_signed.bin
+```
+
```

Comment:
@arturkow2000 lets add here also what may come in next blogposts, like using mainline U-Boot or implementing verified boot

Reply:
Already mentioned in summary.